src/librustc_mir/dataflow/move_paths/builder.rs

```diff
@@ -221,6 +239,22 @@ pub(super) fn gather_moves<'a, 'gcx, 'tcx>(mir: &Mir<'tcx>,
 }
 
 impl<'a, 'gcx, 'tcx> MoveDataBuilder<'a, 'gcx, 'tcx> {
+    fn gather_args(&mut self) {
+        for arg in self.mir.args_iter() {
+            let path = self.data.rev_lookup.locals[arg];
+            let span = self.mir.local_decls[arg].source_info.span;
+
+            let init = self.data.inits.push(Init {
```

Comment:
You want to have arguments assigned before the first statement, to avoid edge cases such as
```Rust
fn foo(x: u8) {
    x=2; // this should be an error, even if we translate the arg in-place.
}
```